local_runner.py

```diff
@@ -0,0 +1,53 @@
+import os
+
+import tfx.v1 as tfx
+from absl import logging
+from tfx_helper.local import LocalPipelineHelper
+
+from pipeline import create_pipeline
+
+
+def run() -> None:
+    """Create and run a pipeline locally."""
+    # Read the pipeline artifact directory from environment variable
+    output_dir = os.environ["PIPELINE_OUTPUT"]
+    # Directory for exporting trained model will be a sub-directory
+    # of the pipeline artifact directory.
+    serving_model_dir = os.path.join(output_dir, "serving_model")
```

Comment:
Why not `Path`?

Reply:
`Path` is analyzing the path and cleaning it up. This works well for local filesystem paths, but the paths we can have here can be remote - most often GCS bucket locations. For those cases `Path` mangles the output (see below how one `/` is missing). `os.path` on the other hand is too dumb to notice that it's not a local path and works correctly. All tensorflow APIs are working on `str` instances and `Path`s are not allowed.

```python
>>> from pathlib import Path
>>> import os
>>> directory = 'gs://bucket/folder'
>>> filename = 'file.txt'
>>> os.path.join(directory, filename)
'gs://bucket/folder/file.txt'
>>> Path(directory) / filename
PosixPath('gs:/bucket/folder/file.txt')
>>> str(Path(directory) / filename)
'gs:/bucket/folder/file.txt'
```

Reply:
I thought so.

Have you considered adding a kind of validation of a proper bucket location format? So that we would allow `Path` and `GCSPath`. In the case of the `Path` instance, we can validate its existence locally. On the other hand, that wouldn't be rather good idea with `GCSPath`

Reply:
I have considered validating the paths, but decided against it. There might be cases that can gain support by installing additional python libs (AFAIK by default TFX ships with GCS and local filesystem support, but there might be ways to add S3 support etc.).

The existence of paths is not mandatory. TFX is automatically creating the directories when it tries to save artifacts. I think that for consistency with the underlying TFX library, we should keep using plain `str`.